File: spring-boot-tipps-reactive/pnia/src/main/java/io/github/jlmc/reactive/domain/services/prefixes/PrefixesService.java
package io.github.jlmc.reactive.domain.services.prefixes;

import org.apache.commons.collections4.list.TreeList;
import org.springframework.cache.Cache;
import org.springframework.cache.CacheManager;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

import java.util.concurrent.atomic.AtomicLong;
import java.util.stream.Collectors;

@Service
public class PrefixesService {

    private static final String CACHE_NAME = "PREFIXES_SERVICE_CACHE";

    private final PrefixesConfiguration prefixesConfiguration;
    private final CacheManager cacheManager;

    private int maxPrefixSize = 7;
    private TreeList<PrefixEntry> prefixes = new TreeList<>();


    public PrefixesService(PrefixesConfiguration prefixesConfiguration, CacheManager cacheManager) {
        this.prefixesConfiguration = prefixesConfiguration;
        this.cacheManager = cacheManager;
    }


    public Mono<String> findByPhoneNumber(String phoneNumber) {
        var prefixResult = Mono.fromCallable(() -> findPrefix(phoneNumber));

        return prefixResult;

        /*
        return Flux.fromIterable(prefixes)
                   .filter(it -> it.match(phoneNumber))
                   .take(1)
                   .log("Prefix found for " + phoneNumber)
                   .map(it -> it.prefix)
                   .next()
                   .switchIfEmpty(Mono.empty());

         */
    }

    public String findPrefix(String phoneNumber) {
        String minorPhoneNumber =
                phoneNumber.length() > maxPrefixSize ?
                        phoneNumber.substring(0, maxPrefixSize) :
                        phoneNumber;

        Cache cache = cacheManager.getCache(CACHE_NAME);

        var cacheable = cache.get(minorPhoneNumber);

        if (cacheable != null) {
            return (String) cacheable.get();
        }

        var value =
                prefixes.stream()
                        .filter(i -> i.match(minorPhoneNumber))
                        .findFirst().map(i -> i.prefix)
                        .orElse(null);

        cache.put(minorPhoneNumber, value);

        return value;
    }

    public void loadPrefixes() throws Exception {
        var prefixes =
                prefixesConfiguration.getPathType()
                                     .readLines(prefixesConfiguration.getFilePath())
                                     .stream()
                                     .map(PrefixEntry::new)
                                     .collect(Collectors.toList());

        this.maxPrefixSize =
                prefixes.stream()
                        .map(it -> it.prefix.length())
                        .max(Integer::compareTo)
                        .orElse(13);

        this.prefixes = new TreeList<>(prefixes);
    }

    static class PrefixEntry implements Comparable<PrefixEntry> {
        private final String prefix;
        private final AtomicLong hits = new AtomicLong(0);

        PrefixEntry(String prefix) {
            this.prefix = prefix;
        }

        private void increment() {
            this.hits.incrementAndGet();
        }

        @Override
        public int compareTo(PrefixEntry o) {
            return Long.compare(o.hits.longValue(), this.hits.longValue());
        }

        public boolean match(String txt) {
            if (txt != null && txt.startsWith(this.prefix)) {
                this.increment();
                return true;
            }
            return false;
        }
    }

}

File: pnia-kotlin/src/main/kotlin/io/github/jlmc/pniakotlin/domain/services/prefixes/PrefixesService.kt
package io.github.jlmc.pniakotlin.domain.services.prefixes

import org.slf4j.LoggerFactory
import org.springframework.cache.CacheManager
import org.springframework.stereotype.Service
import reactor.core.publisher.Mono

@Service
class PrefixesService(
    private val prefixesConfiguration: PrefixesConfiguration,
    private var cacheManager: CacheManager
) {

    companion object {
        private const val CACHE_NAME = "PREFIXES_SERVICE_CACHE"
    }

    private val logger = LoggerFactory.getLogger(PrefixesService::class.java)

    private var prefixes: List<String> = emptyList()
    private var maxPrefixSize = 13

    fun loadPrefixes() {
        val readedPrefixes: List<String> = prefixesConfiguration.pathType?.let { pathType ->
            prefixesConfiguration.filePath?.let { filePath ->
                pathType.readLines(filePath)
            }
        }.orEmpty()

        // for caching proposes
        val maxPrefixLength: Int = readedPrefixes.maxOfOrNull { it.length } ?: 13

        this.prefixes = readedPrefixes
        this.maxPrefixSize = maxPrefixLength
    }

    fun findByPhoneNumber(phoneNumber: String): Mono<String?> {
        return Mono.fromCallable { findPrefix(phoneNumber) }
    }

    private fun findPrefix(phoneNumber: String): String? {
        val minorPhoneNumber =
            if (phoneNumber.length > maxPrefixSize) phoneNumber.substring(0, maxPrefixSize) else phoneNumber

        logger.info("Trying to find prefix of the number <$phoneNumber> with the key <$minorPhoneNumber> ...")

        val cache = cacheManager.getCache(CACHE_NAME)

        val cacheable = cache!![minorPhoneNumber]

        if (cacheable != null) {
            val prefix = cacheable.get() as String?

            logger.info("Found <${prefix != null}> prefix in the cache, for the phone number <$phoneNumber> with the key <$minorPhoneNumber>, the prefix is <${prefix}> ")
            return prefix
        }

        val prefix = prefixes.firstOrNull(minorPhoneNumber::startsWith)

        cache.put(minorPhoneNumber, prefix)

        logger.info("Found <${prefix != null}> prefix out of the cache, for the phone number <$phoneNumber> with the key <$minorPhoneNumber>, the prefix is <${prefix}> ")

        return prefix
    }
}
